dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,467 @@
+	DeviceClassName *string `json:"deviceClassName,omitempty"`
+
+	// Config defines configuration parameters that apply to the requested resource(s).
+	// They are ignored while allocating the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty" protobuf:"bytes,1,opt,name=config"`
 
 	// AdminAccess indicates that this is a claim for administrative access
-	// to the devices. Claims with AdminAccess are expected to be used for
+	// to the device(s). Claims with AdminAccess are expected to be used for
 	// monitoring or other management services for a device.  They ignore
 	// all ordinary claims to the device with respect to access modes and
-	// any resource allocations. Ability to create these claims is
-	// controlled via ResourceQuota.
+	// any resource allocations. Ability to request this kind of access is
+	// controlled via ResourceQuota in the resource.k8s.io API.
 	//
-	// Default is false. If true, a DeviceClass must be specified so that
-	// the Driver is known, and only the Constraints fields will be taken
-	// into account for device selection. All devices meeting the
-	// Constraints expressions will be made available as part of the claim
-	// and may be assigned to a container.
+	// Can be combined with a range to ask for access to all devices
+	// on a node which match the requrirements.
 	//
-	// NOTE: This cannot appear in class because the quota code does not do
-	// an indirection. Still searching for a better way to handle this use
-	// case, without creating a new top-level type for it.
+	// Default is false.
 	//
 	// +optional
 	AdminAccess *bool `json:"adminAccess,omitempty"`
 
-	// Constraints is a CEL expression that operates on device attributes.
-	// In order for a device to be considered, this CEL expression and the
-	// Constraints expression from the DeviceClass must both be true.
+	// FUTURE EXTENSION:
+	// - Count with min, max (including "all that are available").
+	// - Constraints for the set of devices belonging to this request.
+
+	// Requirements describe additional contraints that all must be met by a device
+	// to satisfy the request.
 	//
 	// +optional
-	Constraints *string `json:"constraints,omitempty"`
-
-	// Device claims may be satisfied by choosing multiple devices instead
-	// of just a single device.
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
+}
 
-	// MaxDevices allows the user to specify a maximum number of devices
-	// that may be allocated to sastisfy this claim. The default is equal
-	// to requests.devices. Thus, without specifying MaxDevices, the user
-	// will get exactly the number of devices specified in
-	// requests.devices.
-	//
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the result of that was.
+type ResourceClaimStatus struct {
+	// Allocation is set once the claim has been allocated successfully.
```

Comment:
We need to differentiate between "successfully allocated by the scheduler" and "successfully actuated by driver". I assume this is the "scheduler" version. And yes, it would need to be backed out if a Pod is partially scheduled and then there is a failure in the scheduler, I believe. At least, that is safest.

Reply:
On "successfully actuated by the driver", we have discussed reporting that here. However Sergey has a proposal to report device status in the pod status. That may surface as a separate KEP. Suggest we defer the discussion...